src/main/java/org/apache/sysml/runtime/controlprogram/paramserv/PSWorker.java

```diff
@@ -66,14 +65,10 @@ protected PSWorker(int workerID, String updFunc, Statement.PSFrequency freq, int
 		_ps = ps;
 
 		// Get the update function
-		String[] keys = DMLProgram.splitFunctionKey(updFunc);
-		String funcName = keys[0];
-		String funcNS = null;
-		if (keys.length == 2) {
-			funcNS = keys[0];
-			funcName = keys[1];
-		}
-		FunctionProgramBlock func = ec.getProgram().getFunctionProgramBlock(funcNS, UPDATE_FUNC_PREFIX + _workerID + "_" + funcName);
+		String[] cfn = ParamservUtils.getCompleteFuncName(updFunc, UPDATE_FUNC_PREFIX + workerID + "_");
```

Comment:
I think this naming is the source of the misunderstanding. When we talked about the "update" function I was not referring to the "gradient" function per worker (currently called update function), but the "aggregate" function of the parameter server (which we call "model update time" in the statistics output). For the update strategy per EPOCH, all workers share the same "aggregate" function. The recompiler synchronizes any function or DAG recompilation on the program blocks and DAGs respectively in order to ensure isolation for all scenarios. I would recommend to give each worker (at least for updates per EPOCH) a separate update (i.e., aggregate) function.